src/resources/views/crud/fields/select2_from_ajax.blade.php

```diff
@@ -58,115 +57,152 @@
-@push('crud_fields_scripts')
-<script>
-    function bpFieldInitSelect2FromAjaxElement(element) {
-        var form = element.closest('form');
-        var $placeholder = element.attr('data-placeholder');
-        var $minimumInputLength = element.attr('data-minimum-input-length');
-        var $dataSource = element.attr('data-data-source');
-        var $method = element.attr('data-method');
-        var $fieldAttribute = element.attr('data-field-attribute');
-        var $connectedEntityKeyName = element.attr('data-connected-entity-key-name');
-        var $includeAllFormFields = element.attr('data-include-all-form-fields')=='false' ? false : true;
-        var $allowClear = element.attr('data-column-nullable') == 'true' ? true : false;
-        var $dependencies = JSON.parse(element.attr('data-dependencies'));
-        var $ajaxDelay = element.attr('data-ajax-delay');
-
-        if (!$(element).hasClass("select2-hidden-accessible"))
-        {
-            $(element).select2({
-                theme: 'bootstrap',
-                multiple: false,
-                placeholder: $placeholder,
-                minimumInputLength: $minimumInputLength,
-                allowClear: $allowClear,
-                ajax: {
-                    url: $dataSource,
-                    type: $method,
-                    dataType: 'json',
-                    delay: $ajaxDelay,
-                    data: function (params) {
-                        if ($includeAllFormFields) {
-                            return {
-                                q: params.term, // search term
-                                page: params.page, // pagination
-                                form: form.serializeArray() // all other form inputs
-                            };
-                        } else {
-                            return {
-                                q: params.term, // search term
-                                page: params.page, // pagination
-                            };
-                        }
-                    },
-                    processResults: function (data, params) {
-                        params.page = params.page || 1;
-
-                        var result = {
-                            results: $.map(data.data, function (item) {
-                                textField = $fieldAttribute;
-                                return {
-                                    text: item[textField],
-                                    id: item[$connectedEntityKeyName]
+        <!-- select2_from_ajax field type js -->
+        @loadJsOnce('packages/select2/dist/js/select2.full.min.js')
+        @if (app()->getLocale() !== 'en')
+            @loadJsOnce('packages/select2/dist/js/i18n/' . app()->getLocale() . '.js')
+        @endif
+        @loadOnce('bpFieldInitSelect2FromAjaxElement')
+        <script>
+            if (!window.fetchSingleEntry) {
+                var fetchSingleEntry = function (element) {
```

Comment:
Let's name ```fetchSingleEntry()``` to something more specific, since it's possible that other select2 fields will have a similar function. Maybe something like... ```bpSelect2FromAjaxFetchSingleEntry()```? Ugly but... unique 🤣

Reply:
Or... do we need it as a function at all? If it's only called inside ```bpFieldInitSelect2FromAjaxElement()``` maybe it's better to just leave it inside it. Bonus - we wouldn't need to re-determine all these variables. Right?

Reply:
Alternatively, if we'd want to make it cleaner, we could define it as a custom event on the ```element```, inside ```bpFieldInitSelect2FromAjaxElement()```. Something like:

```js
function bpFieldInitSelect2FromAjaxElement(element) {
    // vars
    // quit if already initialised
    // add select2 on the element
    // if a value already selected, fetch it using an AJAX call
        // bla-bla
        element.trigger('fetch-selected')

    // events
    element.on("fetch-selected", element, function() {
       // actually do the ajax and stuff
    });
}
```

Reply:
I just used promises so they are non-blocking while executing. I used it in other places, and as it probably will be available in other fields (like select2_from_ajax_multiple) i made it a function so i only include it once. 

If you happen to have this two fields you will endup with twice the same code. That's why a function there.

Let me know if I am missing something.

Reply:
I totally agree with using promises. I noticed and appreciated it 👍 

But I don't agree with having it with the same name, in order not to be loaded by each field type. That spells trouble. What if one function gets a little different, for example for ```select2_from_ajax_multiple```? Then it won't get loaded because another one with the same name has already been. Right?

Any way we put it, this won't be DRY. The fact that it will be the same code in multiple files is bad 😄 But it's something that needs to be done at the moment. **Reusing the same name**, so it's not loaded a bunch of times (per field type per page) **is not a solution in my opinion** - the code will still be duplicated across files. **Better to just know each field has all its logic, separated.** It still won't be DRY, but at least it'll be encapsulated.